### PR TITLE
Lidarlite pwm input functionnality

### PR DESCRIPTION
**Pull request by mgirlando**
I wrote a modification to add the Lidar Lite V2 (not tested on V1 but should work) on pwm input of the sonar.
You just plug the device like this :
lidar_lite_pwm_wiring2

The sofware automaticaly detects that the sensor is Lidar and not HCSR04.
No need to add features.
Maybe the name of driver (sonar_hcsr04.*) should be renamed to reflect that.

### CODE REVIEW

**Comment by mgirlando** [comment]
you need to add those stub in  altitude_hold_unittest.cc to pass the test:

```
int32_t sonarGetMaxDistance(void)
{return 400;}

int32_t sonarGetTransitionDistance(void);
{return 300;}
```

File: src/main/drivers/sonar_hcsr04.c [hunk 1]
```diff
@@ -36,15 +35,25 @@
...
+ * The distance between the unit and the object is calculated
+ * by measuring the traveling time of light and output it as the width of a TTL pulse
+ * of 10�s per centimeter.
+ * *** Warning: Lidar Lite V2  operates at +5V (4.5v to 6v)***
+ */
 
 static uint32_t lastMeasurementAt;
 static volatile int32_t measurement = -1;
 static sonarHardware_t const *sonarHardware;
+static bool SonarIsLidarLite = false;
```
Comment (thenickdude) [inline src/main/drivers/sonar_hcsr04.c]: Variable name should start with a lowercase letter.

File: src/main/drivers/sonar_hcsr04.c [hunk 2]
```diff
@@ -52,6 +61,27 @@ static void ECHO_EXTI_IRQHandler(void)
...
         if (timing_stop > timing_start) {
             measurement = timing_stop - timing_start;
+
+            // there are some spurious values arriving but always almost
+            // 1ms over or under (comes from LidarLite when has false detections)
+            // we detect them and remove the suspect ones (not dangerous as they
+            // are always surrounded by two good values)
+            // we detect them and remove the suspect ones
+
+            if(SonarIsLidarLite == true) {
```
Comment (thenickdude) [inline src/main/drivers/sonar_hcsr04.c]: should just be `if (sonarIsLidarLite) {`

File: src/main/drivers/sonar_hcsr04.c [hunk 3]
```diff
@@ -128,12 +158,24 @@ void hcsr04_init(const sonarHardware_t *initialSonarHardware)
...
+        SonarIsLidarLite = true;
+    }
+
+    lastMeasurementAt = millis() - 60; // force 1st measurement in hcsr04_start_reading
 }
 
 // measurement reading is done asynchronously, using interrupt
 void hcsr04_start_reading(void)
 {
+    //if(SonarIsLidarLite == true) return; // not used for Lidar
```
Comment (thenickdude) [inline src/main/drivers/sonar_hcsr04.c]: Line should be uncommented or removed entirely.

File: src/main/drivers/sonar_hcsr04.c [hunk 4]
```diff
@@ -156,17 +199,46 @@ void hcsr04_start_reading(void)
...
 {
-    // The speed of sound is 340 m/s or approx. 29 microseconds per centimeter.
-    // The ping travels out and back, so to find the distance of the
-    // object we take half of the distance traveled.
-    //
-    // 340 m/s = 0.034 cm/microsecond = 29.41176471 *2 = 58.82352941 rounded to 59
-    int32_t distance = measurement / 59;
+    int32_t distance;
+
+    if(SonarIsLidarLite == false) {
```
Comment (thenickdude) [inline src/main/drivers/sonar_hcsr04.c]: Should be `if (!sonarIsLidarLite) {`

**Comment by thenickdude** [comment]
There seems to be very little overlap between the code you're currently combining for the Sonar and Lidar drivers, they should probably just be separated out into separate drivers.

File: src/main/sensors/sonar.c [hunk 5]
```diff
@@ -142,4 +142,19 @@ int32_t sonarGetLatestAltitude(void)
     return calculatedAltitude;
 }
 
+/**
+ * returns the maximum distance readable by the sonar
```
Comment (thenickdude) [inline src/main/sensors/sonar.c]: Would be nice to note units of the return type here (I believe it's centimetres).

**Comment by mgirlando** [comment]
Made all the corrections you requested (coding rules)
Renamed sonar_hcsr04 to pwm_rangers  --> splitting it in 2 drivers would be more complicated as they share same interrupt. (I am creating one to use I2C input to free softserial)
Modified also the unit test (altitude_hold_unittest.cc) but I cannot compile the tests ( hope it works)

**Pull request closed without merging**